tests/smoke_tests/test_basic.py

```diff
@@ -809,3 +811,165 @@ def test_cli_exit_codes(generic_cloud: str):
+                            len(rule.get('port_range')) == 2 and
+                            rule.get('port_range')[0] == test_port_int and
+                            rule.get('port_range')[1] == test_port_int):
+
+                        # Additional check: ensure it's our auto-generated rule
+                        description = rule.get('description', '')
+                        if 'SkyPilot auto-generated' in description and f'port {test_port_int}' in description:
+                            rules_to_remove.append(rule)
+                            print(
+                                f"Found test firewall rule to clean up: TCP {test_port_int}-{test_port_int}"
+                            )
+
+                if rules_to_remove:
+                    print(
+                        f"Cleaning up {len(rules_to_remove)} test firewall rule(s)"
+                    )
+
+                    # Build rule list without our test rules
+                    api_rules = []
+                    for rule in current_rules:
+                        # Check if this rule should be removed
+                        should_remove = False
+                        for rule_to_remove in rules_to_remove:
+                            if (rule.get('protocol')
+                                    == rule_to_remove.get('protocol') and
+                                    rule.get('source_network')
+                                    == rule_to_remove.get('source_network') and
+                                    rule.get('port_range')
+                                    == rule_to_remove.get('port_range')):
+                                should_remove = True
+                                break
+
+                        # Skip if this rule should be removed
+                        if should_remove:
+                            continue
+
+                        if rule.get('protocol') and rule.get('source_network'):
+                            api_rule = {
+                                'protocol': rule.get('protocol'),
+                                'source_network': rule.get('source_network'),
+                                'description': rule.get('description', '')
+                            }
+
+                            # Add port_range for non-icmp protocols
+                            if rule.get('protocol') != 'icmp' and rule.get(
+                                    'port_range'):
+                                api_rule['port_range'] = rule.get('port_range')
+
+                            api_rules.append(api_rule)
+
+                    # Update the rules without our test rule(s)
+                    data = json.dumps({'data': api_rules})
+                    lambda_utils._try_request_with_backoff(
+                        'put',
+                        f'{lambda_utils.API_ENDPOINT}/firewall-rules',
+                        data=data,
+                        headers=lambda_client.headers,
+                    )
```

Comment:
side comment: wow Lambda really needs to add a `DELETE` function instead of clients needing to fetch and update all rules...